### PR TITLE
Feature/maintain skill table

### PR DESCRIPTION
**Pull request by adriaanbakker-java**
extra endpoints for skill table

### CODE REVIEW

File: docs/ENDPOINTS.md [hunk 1]
```diff
@@ -57,4 +58,32 @@ To delete a vacancy. **Returns**:
...
+### POST
+**Requires** 
+a JSON body containing the DTO object of a skill
+see /saveskill
+
+
+## /relinkskills
+### GET
+Deletes links between all vacancies and skills
+Matches all skills with vacancies and creates new links
```
Comment (mostafa-221) [inline docs/ENDPOINTS.md]: This is to update existing resources, so should be done with a PUT request.

File: docs/ENDPOINTS.md [hunk 2]
```diff
@@ -57,4 +58,32 @@ To delete a vacancy. **Returns**:
...
+  'name':'Angular'
+} 
+the id is not yet used at this point, for our purpose, the id is generated in the back end
+
+## /deleteskill
+skill with the corresponding skill name will be deleted, including the links to vacancies
+### POST
+**Requires** 
+a JSON body containing the DTO object of a skill
+see /saveskill
```
Comment (mostafa-221) [inline docs/ENDPOINTS.md:83]: These already exist in the current API

File: src/main/java/nl/ordina/jobcrawler/controller/ResponseCode.java [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+import lombok.Getter;
+
+@Getter
+public class ResponseCode {
+    private String ErrorCode;
+
+    public ResponseCode(String errorCode) {
+        ErrorCode = errorCode;
+    }
+}
```
Comment (mostafa-221) [inline src/main/java/nl/ordina/jobcrawler/controller/ResponseCode.java:12]: From what I gather from the usages, this class is used to create exceptions. Is that true? if yes why not creating the exceptions as their own class?

File: src/main/java/nl/ordina/jobcrawler/controller/SkillController.java [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+@CrossOrigin
+@RestController
+public class SkillController {
+
+
+    private final SkillService skillService;
+
+    private final MatchSkillsService matchSkillsService;
+
+    private final SkillRepository skillRepository;
```
Comment (mostafa-221) [inline src/main/java/nl/ordina/jobcrawler/controller/SkillController.java:31]: Controller layer should not communicate with the repository layer.

File: src/main/java/nl/ordina/jobcrawler/controller/SkillController.java [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+            skillRepository.deleteSkillByName(skillDTO.getName());
+//            skillService.deleteSkill(skillDTO.getName());
+            return new ResponseEntity<>(new ResponseCode("OK"), HttpStatus.OK);
+        } catch (Exception e) {
+            return new ResponseEntity<>(
+                    new ResponseCode("Could not delete skill:" + e.getLocalizedMessage()), HttpStatus.OK);
+        }
+    }
+
+    // getting all skills from database
```
Comment (mostafa-221) [inline src/main/java/nl/ordina/jobcrawler/controller/SkillController.java:109]: This comment is not describing the method below it

File: src/main/java/nl/ordina/jobcrawler/controller/SkillController.java [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+        try {
+            matchSkillsService.relinkSkills();
+            log.info("relink done");
+            return new ResponseEntity<>(new ResponseCode("OK"), HttpStatus.OK);
+        } catch (Exception e) {
+            return new ResponseEntity<>(
+                    new ResponseCode("Could not relink skills:" + e.getLocalizedMessage()), HttpStatus.OK);
+        }
+
+    }
```
Comment (mostafa-221) [inline src/main/java/nl/ordina/jobcrawler/controller/SkillController.java]: This can better go in its own class, take a look at the scraper controller.

File: docs/ENDPOINTS.md [hunk 7]
```diff
@@ -57,4 +58,32 @@ To delete a vacancy. **Returns**:
...
+
+
+## /relinkskills
```
Comment (mostafa-221) [inline docs/ENDPOINTS.md]: this is not a standard RESTful name, as they should not be verbs, but nouns. 
Take a look at these 3 links:
https://spring.io/guides/tutorials/rest/
https://www.springboottutorial.com/spring-boot-crud-rest-service-with-jpa-hibernate
https://bezkoder.com/spring-boot-jpa-crud-rest-api/

File: src/main/java/nl/ordina/jobcrawler/model/Skill.java [hunk 8]
```diff
@@ -19,7 +20,9 @@
 import java.util.UUID;
 
 @Entity
-@Data
+//@Data
+@Getter
+@Setter
```
Comment (mostafa-221) [inline src/main/java/nl/ordina/jobcrawler/model/Skill.java]: why this change?

File: src/main/java/nl/ordina/jobcrawler/model/Skill.java [hunk 9]
```diff
@@ -34,7 +37,7 @@
     // Unique set to false for MVP. Set to true in combination with entitymanager if skill column is still needed at some point.
     private String name;
 
-    @ManyToMany(mappedBy = "skills")
+    @ManyToMany(mappedBy = "skills", fetch = FetchType.EAGER)
```
Comment (mostafa-221) [inline src/main/java/nl/ordina/jobcrawler/model/Skill.java:37]: Eager fetch type should be avoided as much as possible, is there a reason for this one?

File: src/main/java/nl/ordina/jobcrawler/repo/SkillRepository.java [hunk 10]
```diff
@@ -1,22 +1,61 @@
...
 
+    Optional<Skill> findByName(String name);    // Spring makes the query automatically
+
+    @Query(value = "SELECT COUNT(*) FROM vacancy_skills vs WHERE vs.skill_id=?1", nativeQuery = true)
+    int countRelationsById(UUID id);
+
+    @Query(value = "DELETE FROM vacancy_skills vs WHERE vs.vacancy_id=?2 AND skill_id=?1", nativeQuery = true)
+    @Transactional
+    @Modifying
+    void removeRelationsById(UUID skillID, UUID jobID);
```
Comment (mostafa-221) [inline src/main/java/nl/ordina/jobcrawler/repo/SkillRepository.java]: These methods as not used, so please delete them.

File: src/main/java/nl/ordina/jobcrawler/repo/SkillRepository.java [hunk 11]
```diff
@@ -1,22 +1,61 @@
...
+    @Transactional
+    @Modifying
+    void deleteSkills();
+
+
+    @Query(value = "insert into vacancy_skills(vacancy_id, skill_id) values (?1, ?2)",
+            nativeQuery = true)
+    @Transactional
+    @Modifying
+    void linkSkillToVacancy(UUID vacancyId, UUID skillID);
```
Comment (mostafa-221) [inline src/main/java/nl/ordina/jobcrawler/repo/SkillRepository.java]: These are also not used

File: src/main/java/nl/ordina/jobcrawler/service/MatchSkillsService.java [hunk 12]
```diff
@@ -27,14 +27,16 @@
 public class MatchSkillsService {
```
Comment (mostafa-221) [inline src/main/java/nl/ordina/jobcrawler/service/MatchSkillsService.java]: I think a better name would be SkillMatcherService

File: src/main/java/nl/ordina/jobcrawler/service/MatchSkillsService.java [hunk 13]
```diff
@@ -55,9 +57,22 @@ public void changeMatch(Vacancy vacancy) {
...
+    public void relinkSkills() throws Exception {
+            skillService.deleteReferencesToSkills();
+            List<Vacancy> vacancies = vacancyService.findAll();
+            for (Vacancy vacancy: vacancies) {
+                Set<Skill> matchingSkills = findMatchingSkills(vacancy);
+                // not very fast -- could be improved
+                for(Skill skill : matchingSkills) {
+                    skillService.update(vacancy.getId(), skill);
+                }
+//                skillService.createMatchingSkillLinks(vacancy, matchingSkills);
```
Comment (mostafa-221) [inline src/main/java/nl/ordina/jobcrawler/service/MatchSkillsService.java]: commented code can be deleted

File: src/main/java/nl/ordina/jobcrawler/controller/SkillController.java (hunk 4 above)
Comment (mostafa-221) [inline src/main/java/nl/ordina/jobcrawler/controller/SkillController.java]: Controller layer should not communicate with the repository layer directly

File: src/main/java/nl/ordina/jobcrawler/controller/SkillController.java [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+            return new ResponseEntity<>(new ResponseCode("OK"), HttpStatus.OK);
+        } catch (Exception e) {
+            return new ResponseEntity<>(
+                    new ResponseCode("Could not delete skill:" + e.getLocalizedMessage()), HttpStatus.OK);
+        }
+    }
+
+    // getting all skills from database
+    @GetMapping(path = "relinkskills")
+    public ResponseEntity<ResponseCode> relinkSkills() {
```
Comment (mostafa-221) [inline src/main/java/nl/ordina/jobcrawler/controller/SkillController.java]: this comment does not describe the method below it

File: docs/ENDPOINTS.md (hunk 7 above)
Comment (mostafa-221) [reply docs/ENDPOINTS.md]: I think /SkillMatcher would be good

File: src/main/java/nl/ordina/jobcrawler/service/MatchSkillsService.java [hunk 15]
```diff
@@ -55,9 +57,22 @@ public void changeMatch(Vacancy vacancy) {
...
+            }
+    }
 
 
     private void addStandardSkill(String aSkill) {
```
Comment (mostafa-221) [inline src/main/java/nl/ordina/jobcrawler/service/MatchSkillsService.java]: This is for testing purposes, so it can be deleted or moved to test classes

File: src/main/java/nl/ordina/jobcrawler/service/MatchSkillsService.java [hunk 16]
```diff
@@ -69,9 +84,9 @@ private void addStandardSkill(String aSkill) {
     public void insertStandardSkills() {
```
Comment (mostafa-221) [inline src/main/java/nl/ordina/jobcrawler/service/MatchSkillsService.java]: This method is not used. Furthermore, it can be better made into an sql file in the test resources.

File: src/main/java/nl/ordina/jobcrawler/service/SkillService.java [hunk 17]
```diff
@@ -94,4 +94,15 @@ public boolean delete(UUID id) {
...
+        skillRepository.deleteReferencesToSkills();
+    }
+
+    public void deleteReferencesToSkill(String name) {
+        skillRepository.deleteReferencesToSkill(name);
+    }
+
+    public void deleteAll() {
+        skillRepository.deleteAll();
+    }
```
Comment (mostafa-221) [inline src/main/java/nl/ordina/jobcrawler/service/SkillService.java]: This is for testing purposes, so it can be put in the tests

File: src/main/java/nl/ordina/jobcrawler/service/SkillService.java [hunk 18]
```diff
@@ -94,4 +94,15 @@ public boolean delete(UUID id) {
     }
 
 
+    public void deleteReferencesToSkills() {
+        skillRepository.deleteReferencesToSkills();
+    }
+
+    public void deleteReferencesToSkill(String name) {
+        skillRepository.deleteReferencesToSkill(name);
+    }
```
Comment (mostafa-221) [inline src/main/java/nl/ordina/jobcrawler/service/SkillService.java:109]: maybe it is a good idea to only have one skill and if the caller needs to delete a set, then it has to implement a loop

**Comment by mostafa-221** [comment]
I do not see any unit tests. can you please implement them?

File: docs/ENDPOINTS.md (hunk 2 above)
Comment (adriaanbakker-java) [reply docs/ENDPOINTS.md:83]: no, did not work with the DTO object

File: src/main/java/nl/ordina/jobcrawler/controller/ResponseCode.java (hunk 3 above)
Comment (adriaanbakker-java) [reply src/main/java/nl/ordina/jobcrawler/controller/ResponseCode.java:12]: no, it is to create an error message

File: docs/ENDPOINTS.md (hunk 1 above)
Comment (adriaanbakker-java) [reply docs/ENDPOINTS.md:89]: ok, will adapt

File: src/main/java/nl/ordina/jobcrawler/controller/SkillController.java (hunk 6 above)
Comment (adriaanbakker-java) [reply src/main/java/nl/ordina/jobcrawler/controller/SkillController.java]: ok, will take a look at it

File: docs/ENDPOINTS.md (hunk 7 above)
Comment (adriaanbakker-java) [reply docs/ENDPOINTS.md]: Ok, as long as it does not suggest CRUD operation because it isn't, it is a method to reconcile relationships

File: src/main/java/nl/ordina/jobcrawler/repo/SkillRepository.java (hunk 10 above)
Comment (adriaanbakker-java) [reply src/main/java/nl/ordina/jobcrawler/repo/SkillRepository.java]: deleted

File: src/main/java/nl/ordina/jobcrawler/repo/SkillRepository.java (hunk 11 above)
Comment (adriaanbakker-java) [reply src/main/java/nl/ordina/jobcrawler/repo/SkillRepository.java]: for the moment the deleteSkills operation and linkSkillToVacancy operation are used for development purposes 
moves to sql file

File: src/main/java/nl/ordina/jobcrawler/service/MatchSkillsService.java (hunk 12 above)
Comment (adriaanbakker-java) [reply src/main/java/nl/ordina/jobcrawler/service/MatchSkillsService.java]: Adapted

File: src/main/java/nl/ordina/jobcrawler/service/MatchSkillsService.java (hunk 13 above)
Comment (adriaanbakker-java) [reply src/main/java/nl/ordina/jobcrawler/service/MatchSkillsService.java]: adapted

File: src/main/java/nl/ordina/jobcrawler/controller/SkillController.java (hunk 4 above)
Comment (adriaanbakker-java) [reply src/main/java/nl/ordina/jobcrawler/controller/SkillController.java]: adapted

File: src/main/java/nl/ordina/jobcrawler/controller/SkillController.java (hunk 14 above)
Comment (adriaanbakker-java) [reply src/main/java/nl/ordina/jobcrawler/controller/SkillController.java]: adapted, method is now moved to MatchSkillController and is the /skillmatcher endpoint method

File: src/main/java/nl/ordina/jobcrawler/service/MatchSkillsService.java (hunk 15 above)
Comment (adriaanbakker-java) [reply src/main/java/nl/ordina/jobcrawler/service/MatchSkillsService.java]: will. move to sql file in future release, for now deleted

File: src/main/java/nl/ordina/jobcrawler/service/MatchSkillsService.java (hunk 16 above)
Comment (adriaanbakker-java) [reply src/main/java/nl/ordina/jobcrawler/service/MatchSkillsService.java]: will move to sql file in future, for now deleted

File: src/main/java/nl/ordina/jobcrawler/service/SkillService.java (hunk 17 above)
Comment (adriaanbakker-java) [reply src/main/java/nl/ordina/jobcrawler/service/SkillService.java]: will move to sql file in future, for now deleted

File: src/main/java/nl/ordina/jobcrawler/service/SkillService.java (hunk 18 above)
Comment (adriaanbakker-java) [reply src/main/java/nl/ordina/jobcrawler/service/SkillService.java:109]: no, will create extra code that has more logic in it

File: docs/ENDPOINTS.md (hunk 1 above)
Comment (adriaanbakker-java) [reply docs/ENDPOINTS.md]: adapted

File: src/main/java/nl/ordina/jobcrawler/controller/SkillController.java (hunk 6 above)
Comment (adriaanbakker-java) [reply src/main/java/nl/ordina/jobcrawler/controller/SkillController.java]: separate service MatchSkillService

File: docs/ENDPOINTS.md (hunk 7 above)
Comment (adriaanbakker-java) [reply docs/ENDPOINTS.md]: /skillmatcher because all our endpoints are lowercase only as until now

File: docs/ENDPOINTS.md (hunk 1 above)
Comment (adriaanbakker-java) [reply docs/ENDPOINTS.md]: adapted

File: docs/ENDPOINTS.md (hunk 2 above)
Comment (mostafa-221) [reply docs/ENDPOINTS.md:83]: The saving of a skill can be done by making a POST request with the body to /skills and the delete can be done with a DELETE request to /skills/{id} where id is the id of the skill.

File: src/main/java/nl/ordina/jobcrawler/controller/ResponseCode.java (hunk 3 above)
Comment (mostafa-221) [reply src/main/java/nl/ordina/jobcrawler/controller/ResponseCode.java:12]: But why not creating custom exceptions? this way it is also readable

File: src/main/java/nl/ordina/jobcrawler/service/ScraperService.java [hunk 19]
```diff
@@ -27,13 +26,16 @@
 @Service
 public class ScraperService {
 
-    private final VacancyService vacancyService;
-    private final MatchSkillsService matchSkillsService;
+
+    private VacancyService vacancyService;
+
+    private SkillMatcherService skillMatcherService;
```
Comment (mostafa-221) [inline src/main/java/nl/ordina/jobcrawler/service/ScraperService.java]: why not final?

File: src/main/java/nl/ordina/jobcrawler/service/SkillService.java (hunk 17 above)
Comment (mostafa-221) [reply src/main/java/nl/ordina/jobcrawler/service/SkillService.java]: is the sql file following in a future commit?

File: src/main/java/nl/ordina/jobcrawler/service/ScraperService.java (hunk 19 above)
Comment (adriaanbakker-java) [reply src/main/java/nl/ordina/jobcrawler/service/ScraperService.java]: adapted

**Review by narayan190290 — APPROVED** [review]
